sdks/java/io/tika/src/main/java/org/apache/beam/sdk/io/tika/TikaSource.java

```diff
@@ -0,0 +1,466 @@
+
+      if (spec.getMinimumTextLength() != null) {
+        tikaHandler.setMinTextLength(spec.getMinimumTextLength());
+      }
+
+      if (!Boolean.TRUE.equals(spec.getParseSynchronously())) {
+        // Try to parse the file on the executor thread to make the best effort
+        // at letting the pipeline thread advancing over the file content
+        // without immediately parsing all of it
+        execService = Executors.newFixedThreadPool(1);
+        execService.submit(new Runnable() {
+          public void run() {
+            try {
+              parser.parse(is, tikaHandler, tikaMetadata, context);
+              is.close();
+            } catch (Exception ex) {
+              tikaHandler.setParseException(ex);
+            }
+          }
+        });
+      } else {
+        // Some parsers might not be able to report the content in chunks.
+        // It does not make sense to create extra threads in such cases
+        try {
+          parser.parse(is, tikaHandler, tikaMetadata, context);
+        } catch (Exception ex) {
+          throw new IOException(ex);
+        } finally {
+          is.close();
+        }
+      }
+      return advanceToNext();
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      checkState(current != null, "Call start() before advance()");
+      return advanceToNext();
+    }
+
+    protected boolean advanceToNext() throws IOException {
+      current = null;
+      // The content is reported first
+      if (metadataIterator == null) {
+        // Check if some content is already available
+        current = tikaHandler.getCurrent();
+
+        if (current == null && !Boolean.TRUE.equals(spec.getParseSynchronously())) {
+          long maxPollTime = 0;
+          long configuredMaxPollTime = spec.getQueueMaxPollTime() == null
+              ? TikaIO.Read.DEFAULT_QUEUE_MAX_POLL_TIME : spec.getQueueMaxPollTime();
+          long configuredPollTime = spec.getQueuePollTime() == null
+              ? TikaIO.Read.DEFAULT_QUEUE_POLL_TIME : spec.getQueuePollTime();
+
+          // Poll the queue till the next piece of data is available
+          while (current == null && maxPollTime < configuredMaxPollTime) {
+            boolean docEnded = tikaHandler.waitForNext(configuredPollTime);
+            current = tikaHandler.getCurrent();
+            // End of Document ?
+            if (docEnded) {
+              break;
+            }
+            maxPollTime += spec.getQueuePollTime();
+          }
+        }
+        // No more content ?
+        if (current == null && Boolean.TRUE.equals(spec.getReadOutputMetadata())) {
+          // Time to report the metadata
+          metadataIterator = Arrays.asList(tikaMetadata.names()).iterator();
+        }
+      }
+
+      if (metadataIterator != null && metadataIterator.hasNext()) {
+          String key = metadataIterator.next();
+          // The metadata name/value separator can be configured if needed
```

Comment:
It just should have been a KV: the data you are returning is not a String, it is key-value pairs.
This will also have ambiguous parsing in case the key or the value can have a "=" character in them.

Reply:
No, it is only the metadata which are not reported by default; and the comment there suggests that this character can be configured...